aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +26,236 @@ namespace at {
-  checkBackend("CPU_tensor_apply4", {tensor1, tensor2, tensor3, tensor4}, Backend::CPU);
-  bool TH_TENSOR_APPLY_hasFinished = false;
-  int64_t TH_TENSOR_dim_index = 0;
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar1, tensor1, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar2, tensor2, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar3, tensor3, dim, 1)
-  __ATH_TENSOR_APPLYX_PREAMBLE(scalar4, tensor4, dim, 1)
+inline bool continue_loop() {
+  return true;
+}
 
-  int elements_equal = 1;
-  auto t1_numel = tensor1.numel();
-  auto t2_numel = tensor2.numel();
-  auto t3_numel = tensor3.numel();
-  auto t4_numel = tensor4.numel();
-  if(t1_numel!= t2_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t3_numel) {
-    elements_equal = 0;
-  } else if(t1_numel != t4_numel) {
-      elements_equal = 0;
-  }
-  if (elements_equal == 0) {
-    std::ostringstream oss;
-    oss << "inconsistent tensor size, expected " << tensor1.sizes() << ", " << tensor2.sizes() << ", "
-        << tensor3.sizes() << ", and " << tensor4.sizes() << " to have the same number of elements, but got "
-        << t1_numel << ", " << t2_numel << ", " << t3_numel << ", and " << t4_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+template <typename Arg, typename... Args>
+inline bool continue_loop(Arg& iter, Args&... iter_tail) {
+  bool result = false;
+  if (iter.dim_ != 0) {
+    result = (iter.counter_[iter.dim_ - 1] < iter.sizes_[iter.dim_ - 1]);
   }
+  return result && continue_loop(iter_tail...);
+}
+
+inline void increase_counters(){};
 
-  while(!TH_TENSOR_APPLY_hasFinished)
-  {
-    /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i <  tensor1_size && tensor2_i < tensor2_size && tensor3_i < tensor3_size && tensor4_i < tensor4_size
-        ; tensor1_i++, tensor2_i++, tensor3_i++, tensor4_i++,
-          tensor1_data += tensor1_stride, tensor2_data += tensor2_stride, tensor3_data += tensor3_stride, tensor4_data += tensor4_stride)
-    {
-      op(*tensor1_data, *tensor2_data, *tensor3_data, *tensor4_data);
+template <typename Arg, typename... Args>
+inline void increase_counters(Arg& iter, Args&... iter_tail) {
+  iter.counter_[iter.dim_ - 1]++;
+  iter.data_ += iter.strides_[iter.dim_ - 1];
+  increase_counters(iter_tail...);
+}
+
+inline void apply_op(){};
+
+template <typename Op, typename... Args>
+inline void apply_op(int64_t numel, const Op& op, Args... iters) {
+  // For 0-dim tensors
+  if (numel == 1 && !continue_loop(iters...)) {
+    op(*iters.data_...);
+    return;
+  }
+  for (int64_t i = 0; i < numel;) {
+    for (; continue_loop(iters...); increase_counters(iters...)) {
+      op(*iters.data_...);
+      i++;
     }
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor3, 0)
-    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor4, 0)
+    iterate_all(iters...);
```

Comment:
Do I understand correctly that `increase_counters` just blindly increments them, while `iterate` takes care of dim overflows in counters? Why the distinction? It looks like `iterate_all` is always followed by `increase_counters`, so why keep them separate?

Reply:
Good point, I'll merge them.